notifications/postgres/notificator.go

```diff
@@ -0,0 +1,336 @@
+		} else {
+			n.updateLastKnownRevision(payload.Revision)
+			n.processNotificationPayload(payload)
+		}
+	}
+}
+
+func (n *notificator) getPayload(data string) (*notifyEventPayload, error) {
+	payload := &notifyEventPayload{}
+	if err := json.Unmarshal([]byte(data), payload); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func (n *notificator) processNotificationPayload(payload *notifyEventPayload) {
+	notificationPlatformID := payload.PlatformID
+	notificationID := payload.NotificationID
+	recipients := n.getRecipients(notificationPlatformID)
+	if len(recipients) == 0 {
+		return
+	}
+	notification, err := n.storage.GetNotification(n.ctx, notificationID)
+	if err != nil {
+		log.C(n.ctx).WithError(err).Errorf("notification %s could not be retrieved from the DB, closing consumers", notificationID)
+		n.closeAllConsumers()
+		return
+	}
+	for _, platformConsumers := range recipients {
+		n.sendNotificationToPlatformConsumers(platformConsumers, notification)
+	}
+}
+
+func (n *notificator) getRecipients(platformID string) consumers {
+	n.consumersMutex.RLock()
+	defer n.consumersMutex.RUnlock()
+	if platformID == "" {
+		return n.consumers
+	}
+	platformConsumers, found := n.consumers[platformID]
+	if !found {
+		return nil
+	}
+	return consumers{
+		platformID: platformConsumers,
+	}
+}
+
+func (n *notificator) sendNotificationToPlatformConsumers(platformConsumers []notifications.NotificationQueue, notification *types.Notification) {
+	for _, consumer := range platformConsumers {
+		if err := consumer.Enqueue(notification); err != nil {
+			log.C(n.ctx).WithError(err).Infof("consumer %s notification queue returned error %v", consumer.ID(), err)
+			consumer.Close()
+		}
+	}
+}
+
+func (n *notificator) awaitTermination() {
+	<-n.ctx.Done()
+	logger := log.C(n.ctx)
+	logger.Info("context cancelled, stopping notificator...")
+	n.isRunning = false
+	n.stopConnection()
+	n.group.Done()
+}
+
+func (n *notificator) stopConnection() {
+	err := n.stopListening()
+	logger := log.C(n.ctx)
+	if err != nil {
+		logger.WithError(err).Info("could not unlisten notification channel")
+	}
+	n.connectionMutex.Lock()
+	defer n.connectionMutex.Unlock()
+	if err = n.connection.Close(); err != nil {
+		logger.WithError(err).Info("could not close db connection")
+	}
+}
+
+func (n *notificator) stopListening() error {
+	n.connectionMutex.Lock()
+	defer n.connectionMutex.Unlock()
+	if !n.isListening {
+		return nil
+	}
+	n.updateLastKnownRevision(invalidRevisionNumber)
+	err := n.connection.Unlisten(postgresChannel)
+	if err == nil {
+		n.isListening = false
+	}
+	return err
+}
+
+func (n *notificator) startListening() error {
+	n.connectionMutex.Lock()
+	defer n.connectionMutex.Unlock()
+	if n.isListening {
+		return nil
+	}
+	lastKnownRevision, err := n.storage.GetLastRevision(n.ctx)
+	if err != nil {
+		return err
+	}
+	n.updateLastKnownRevision(lastKnownRevision)
+
+	err = n.connection.Listen(postgresChannel)
```

Comment:
`postgresChannel` might be outdated if a notification arrives between `GetLastRevision` and `Listen`

Reply:
refactored